File: 01-programming/go/ejercicios/proyectos/proyecto-final/tests/biblioteca_test.go
package main

import (
	"errors"
	"strings"
	"testing"

	"proyectofinal/datos"
	"proyectofinal/modelos"
	"proyectofinal/servicios"
)

func setup() (*servicios.BibliotecaService, *servicios.ReportesService) {
	libros := datos.Nuevo[modelos.Libro]()
	miembros := datos.Nuevo[modelos.Miembro]()
	prestamos := datos.Nuevo[modelos.Prestamo]()
	s := servicios.NuevoBiblioteca(libros, miembros, prestamos)
	r := servicios.NuevoReportes(s)
	return s, r
}

func TestAltaLibro(t *testing.T) {
	s, _ := setup()

	libro, err := s.AltaLibro("El Quijote", "Miguel de Cervantes", "9788420412146", 1605, modelos.Ficcion)
	if err != nil {
		t.Fatalf("AltaLibro inesperado: %v", err)
	}
	if libro.Titulo != "El Quijote" {
		t.Errorf("título = %q, se esperaba %q", libro.Titulo, "El Quijote")
	}
	if !libro.Disponible {
		t.Errorf("el libro debe crearse disponible")
	}
}

func TestAltaLibroValidaDatos(t *testing.T) {
	s, _ := setup()

	if _, err := s.AltaLibro("", "Autor", "1", 2020, modelos.Otro); !errors.Is(err, servicios.ErrDatosInvalidos) {
		t.Errorf("título vacío debe devolver ErrDatosInvalidos, se obtuvo %v", err)
	}
	if _, err := s.AltaLibro("Título", "", "1", 2020, modelos.Otro); !errors.Is(err, servicios.ErrDatosInvalidos) {
		t.Errorf("autor vacío debe devolver ErrDatosInvalidos, se obtuvo %v", err)
	}
}

func TestBuscarLibros(t *testing.T) {
	s, _ := setup()
	s.AltaLibro("El Quijote", "Miguel de Cervantes", "1", 1605, modelos.Ficcion)
	s.AltaLibro("Cien años de soledad", "Gabriel García Márquez", "2", 1967, modelos.Ficcion)
	s.AltaLibro("Introducción a Go", "Alan Donovan", "3", 2015, modelos.Tecnologia)

	resultados := s.BuscarLibros("quijote")
	if len(resultados) != 1 {
		t.Fatalf("se esperaba 1 resultado, se obtuvo %d", len(resultados))
	}
	if resultados[0].Titulo != "El Quijote" {
		t.Errorf("resultado = %q, se esperaba El Quijote", resultados[0].Titulo)
	}

	resultados = s.BuscarLibros("go")
	if len(resultados) != 1 {
		t.Fatalf("búsqueda 'go' debe dar 1 resultado, se obtuvo %d", len(resultados))
	}

	if s.BuscarLibros("inexistente") == nil {
		t.Errorf("BuscarLibros sin coincidencias debe devolver slice vacío, no nil")
	}
}

func TestAltaMiembroYEmailDuplicado(t *testing.T) {
	s, _ := setup()

	m, err := s.AltaMiembro("Ana López", "ana@correo.com", "600111222")
	if err != nil {
		t.Fatalf("AltaMiembro inesperado: %v", err)
	}
	if !m.Activo {
		t.Errorf("el miembro debe crearse activo")
	}

	if _, err := s.AltaMiembro("Ana 2", "ana@correo.com", "600111223"); !errors.Is(err, servicios.ErrEmailDuplicado) {
		t.Errorf("email repetido debe devolver ErrEmailDuplicado, se obtuvo %v", err)
	}

	if _, err := s.AltaMiembro("", "otro@correo.com", "600111224"); !errors.Is(err, servicios.ErrDatosInvalidos) {
		t.Errorf("nombre vacío debe devolver ErrDatosInvalidos, se obtuvo %v", err)
	}
}

func TestCrearPrestamo(t *testing.T) {
	s, _ := setup()

	libro, _ := s.AltaLibro("El Quijote", "Cervantes", "1", 1605, modelos.Ficcion)
	miembro, _ := s.AltaMiembro("Ana López", "ana@correo.com", "600111222")

	p, err := s.CrearPrestamo(libro.ID, miembro.ID)
	if err != nil {
		t.Fatalf("CrearPrestamo inesperado: %v", err)
	}
	if p.Devuelto {
		t.Errorf("un préstamo nuevo no debe estar devuelto")
	}

	// El mismo libro no puede prestarse de nuevo
	if _, err := s.CrearPrestamo(libro.ID, miembro.ID); !errors.Is(err, servicios.ErrLibroNoDisponible) {
		t.Errorf("libro prestado debe devolver ErrLibroNoDisponible, se obtuvo %v", err)
	}
}

func TestCrearPrestamoLibroNoExiste(t *testing.T) {
	s, _ := setup()
	miembro, _ := s.AltaMiembro("Ana", "ana@correo.com", "1")

	if _, err := s.CrearPrestamo(999, miembro.ID); !errors.Is(err, servicios.ErrEntidadNoEncontrada) {
		t.Errorf("libro inexistente debe devolver ErrEntidadNoEncontrada, se obtuvo %v", err)
	}
}

func TestDevolverPrestamo(t *testing.T) {
	s, _ := setup()

	libro, _ := s.AltaLibro("El Quijote", "Cervantes", "1", 1605, modelos.Ficcion)
	miembro, _ := s.AltaMiembro("Ana", "ana@correo.com", "1")
	p, _ := s.CrearPrestamo(libro.ID, miembro.ID)

	devuelto, err := s.DevolverPrestamo(p.ID)
	if err != nil {
		t.Fatalf("DevolverPrestamo inesperado: %v", err)
	}
	if !devuelto.Devuelto {
		t.Errorf("el préstamo debe quedar devuelto")
	}

	// El libro vuelve a estar disponible
	_, err = s.CrearPrestamo(libro.ID, miembro.ID)
	if err != nil {
		t.Errorf("tras la devolución el libro debe poder prestarse, se obtuvo %v", err)
	}

	if _, err := s.DevolverPrestamo(999); !errors.Is(err, servicios.ErrEntidadNoEncontrada) {
		t.Errorf("préstamo inexistente debe devolver ErrEntidadNoEncontrada, se obtuvo %v", err)
	}
}

func TestResumen(t *testing.T) {
	s, r := setup()

	s.AltaLibro("A", "Autor", "1", 2020, modelos.Otro)
	s.AltaLibro("B", "Autor", "2", 2020, modelos.Otro)
	m, _ := s.AltaMiembro("Ana", "ana@correo.com", "1")
	_, _ = s.CrearPrestamo(1, m.ID)

	resumen := r.Resumen()
	if resumen["libros"] != 2 {
		t.Errorf("libros = %d, se esperaba 2", resumen["libros"])
	}
	if resumen["libros_disponibles"] != 1 {
		t.Errorf("libros_disponibles = %d, se esperaba 1", resumen["libros_disponibles"])
	}
	if resumen["libros_prestados"] != 1 {
		t.Errorf("libros_prestados = %d, se esperaba 1", resumen["libros_prestados"])
	}
	if resumen["miembros_activos"] != 1 {
		t.Errorf("miembros_activos = %d, se esperaba 1", resumen["miembros_activos"])
	}
	if resumen["prestamos_activos"] != 1 {
		t.Errorf("prestamos_activos = %d, se esperaba 1", resumen["prestamos_activos"])
	}
}

func TestTopLibros(t *testing.T) {
	s, r := setup()

	libroA, _ := s.AltaLibro("A", "Autor", "1", 2020, modelos.Otro)
	libroB, _ := s.AltaLibro("B", "Autor", "2", 2020, modelos.Otro)
	m, _ := s.AltaMiembro("Ana", "ana@correo.com", "1")

	_, _ = s.CrearPrestamo(libroA.ID, m.ID)
	_, _ = s.DevolverPrestamo(1)
	_, _ = s.CrearPrestamo(libroA.ID, m.ID)
	_, _ = s.DevolverPrestamo(2)
	_, _ = s.CrearPrestamo(libroB.ID, m.ID)

	top := r.TopLibros(2)
	if len(top) != 2 {
		t.Fatalf("TopLibros debe devolver 2, se obtuvo %d", len(top))
	}
	if top[0].Nombre != "A" || top[0].Prestamos != 2 {
		t.Errorf("top[0] = %+v, se esperaba A con 2 préstamos", top[0])
	}
	if top[1].Nombre != "B" || top[1].Prestamos != 1 {
		t.Errorf("top[1] = %+v, se esperaba B con 1 préstamo", top[1])
	}
}

func TestTopMiembros(t *testing.T) {
	s, r := setup()

	m1, _ := s.AltaMiembro("Ana", "ana@correo.com", "1")
	m2, _ := s.AltaMiembro("Luis", "luis@correo.com", "2")

	libroA, _ := s.AltaLibro("A", "Autor", "1", 2020, modelos.Otro)
	libroB, _ := s.AltaLibro("B", "Autor", "2", 2020, modelos.Otro)

	_, _ = s.CrearPrestamo(libroA.ID, m1.ID)
	_, _ = s.CrearPrestamo(libroB.ID, m1.ID)

	// m2 nunca pide prestado, aparece con 0 préstamos.
	_ = m2

	top := r.TopMiembros(2)
	if len(top) != 2 {
		t.Fatalf("TopMiembros debe devolver 2, se obtuvo %d", len(top))
	}
	if top[0].Nombre != "Ana" || top[0].Prestamos != 2 {
		t.Errorf("top[0] = %+v, se esperaba Ana con 2 préstamos", top[0])
	}
	if top[1].Nombre != "Luis" || top[1].Prestamos != 0 {
		t.Errorf("top[1] = %+v, se esperaba Luis con 0 préstamos", top[1])
	}
}

func TestPrestamosVencidos(t *testing.T) {
	s, _ := setup()
	// No hay préstamos con fechas pasadas recién creados, solo validamos que
	// un préstamo recién creado no aparece como vencido.
	libro, _ := s.AltaLibro("A", "Autor", "1", 2020, modelos.Otro)
	m, _ := s.AltaMiembro("Ana", "ana@correo.com", "1")
	_, _ = s.CrearPrestamo(libro.ID, m.ID)

	if vencidos := s.PrestamosVencidos(); len(vencidos) != 0 {
		t.Errorf("PrestamosVencidos = %d, se esperaba 0", len(vencidos))
	}
}

func TestTopLibrosDescendente(t *testing.T) {
	s, r := setup()
	libroA, _ := s.AltaLibro("A", "Autor", "1", 2020, modelos.Otro)
	m, _ := s.AltaMiembro("Ana", "ana@correo.com", "1")

	_, _ = s.CrearPrestamo(libroA.ID, m.ID)
	_, _ = s.DevolverPrestamo(1)
	_, _ = s.CrearPrestamo(libroA.ID, m.ID)

	top := r.TopLibros(1)
	if len(top) != 1 {
		t.Fatalf("TopLibros(1) debe devolver 1 elemento")
	}
	if top[0].Prestamos != 2 {
		t.Errorf("debe contar préstamos acumulados, se obtuvo %d", top[0].Prestamos)
	}
}

// Comprueba que la búsqueda no distingue mayúsculas.
func TestBuscarSinMayusculas(t *testing.T) {
	s, _ := setup()
	s.AltaLibro("Cien Años de Soledad", "Gabriel García Márquez", "2", 1967, modelos.Ficcion)

	if res := s.BuscarLibros("años"); len(res) != 1 {
		t.Errorf("búsqueda en minúsculas debe encontrar 'Años', se obtuvo %d", len(res))
	}
	if res := s.BuscarLibros("GARCÍA"); len(res) != 1 {
		t.Errorf("búsqueda en mayúsculas debe encontrar 'García', se obtuvo %d", len(res))
	}
}

// Valida que AltaLibro asigne IDs incrementales.
func TestIDsIncrementales(t *testing.T) {
	s, _ := setup()
	libro1, _ := s.AltaLibro("A", "Autor", "1", 2020, modelos.Otro)
	libro2, _ := s.AltaLibro("B", "Autor", "2", 2020, modelos.Otro)

	if libro1.ID != 1 || libro2.ID != 2 {
		t.Errorf("IDs = (%d, %d), se esperaba (1, 2)", libro1.ID, libro2.ID)
	}
}

// Valida que los IDs asignados al crear préstamos no se pisan.
func TestIDsPrestamos(t *testing.T) {
	s, _ := setup()
	libroA, _ := s.AltaLibro("A", "Autor", "1", 2020, modelos.Otro)
	libroB, _ := s.AltaLibro("B", "Autor", "2", 2020, modelos.Otro)
	m, _ := s.AltaMiembro("Ana", "ana@correo.com", "1")

	p1, _ := s.CrearPrestamo(libroA.ID, m.ID)
	p2, _ := s.CrearPrestamo(libroB.ID, m.ID)

	if p1.ID == p2.ID {
		t.Errorf("los préstamos no deben compartir ID (%d)", p1.ID)
	}
}

var _ = strings.TrimSpace

File: 01-programming/rust/ejercicios/proyectos/proyecto-final/tests/biblioteca_test.rs
//! Tests de referencia de la capa de servicios.
//!
//! Ejecuta con `cargo test` desde esta carpeta.

use proyectofinal::datos::Repositorio;
use proyectofinal::modelos::{GeneroLibro, Libro, Miembro, Prestamo};
use proyectofinal::servicios::{BibliotecaService, ErrorBiblioteca, ReportesService};

fn setup() -> BibliotecaService {
    BibliotecaService::nuevo(
        Repositorio::<Libro>::nuevo(),
        Repositorio::<Miembro>::nuevo(),
        Repositorio::<Prestamo>::nuevo(),
    )
}

#[test]
fn alta_libro() {
    let mut s = setup();
    let libro = s
        .alta_libro("El Quijote", "Miguel de Cervantes", "9788420412146", 1605, GeneroLibro::Ficcion)
        .expect("AltaLibro inesperado");
    assert_eq!(libro.titulo, "El Quijote");
    assert!(libro.disponible, "el libro debe crearse disponible");
}

#[test]
fn alta_libro_valida_datos() {
    let mut s = setup();
    let err_titulo = s.alta_libro("", "Autor", "1", 2020, GeneroLibro::Otro);
    assert_eq!(err_titulo, Err(ErrorBiblioteca::DatosInvalidos));
    let err_autor = s.alta_libro("Titulo", "", "1", 2020, GeneroLibro::Otro);
    assert_eq!(err_autor, Err(ErrorBiblioteca::DatosInvalidos));
}

#[test]
fn buscar_libros() {
    let mut s = setup();
    s.alta_libro("El Quijote", "Miguel de Cervantes", "1", 1605, GeneroLibro::Ficcion).unwrap();
    s.alta_libro("Cien años de soledad", "Gabriel García Márquez", "2", 1967, GeneroLibro::Ficcion).unwrap();
    s.alta_libro("Introducción a Rust", "Steve Klabnik", "3", 2015, GeneroLibro::Tecnologia).unwrap();

    let resultados = s.buscar_libros("quijote");
    assert_eq!(resultados.len(), 1);
    assert_eq!(resultados[0].titulo, "El Quijote");

    let resultados = s.buscar_libros("rust");
    assert_eq!(resultados.len(), 1);

    let sin_resultados = s.buscar_libros("inexistente");
    assert!(sin_resultados.is_empty(), "sin coincidencias debe ser Vec vacío");
}

#[test]
fn alta_miembro_y_email_duplicado() {
    let mut s = setup();
    let m = s
        .alta_miembro("Ana López", "ana@correo.com", "600111222")
        .expect("AltaMiembro inesperado");
    assert!(m.activo, "el miembro debe crearse activo");

    let duplicado = s.alta_miembro("Ana 2", "ana@correo.com", "600111223");
    assert_eq!(duplicado, Err(ErrorBiblioteca::EmailDuplicado));

    let invalido = s.alta_miembro("", "otro@correo.com", "600111224");
    assert_eq!(invalido, Err(ErrorBiblioteca::DatosInvalidos));
}

#[test]
fn crear_prestamo() {
    let mut s = setup();
    let libro = s
        .alta_libro("El Quijote", "Cervantes", "1", 1605, GeneroLibro::Ficcion)
        .unwrap();
    let miembro = s.alta_miembro("Ana López", "ana@correo.com", "600111222").unwrap();

    let p = s.crear_prestamo(libro.id, miembro.id).expect("CrearPrestamo inesperado");
    assert!(!p.devuelto, "un préstamo nuevo no debe estar devuelto");

    let repetido = s.crear_prestamo(libro.id, miembro.id);
    assert_eq!(repetido, Err(ErrorBiblioteca::LibroNoDisponible));
}

#[test]
fn crear_prestamo_libro_no_existe() {
    let mut s = setup();
    let miembro = s.alta_miembro("Ana", "ana@correo.com", "1").unwrap();
    let err = s.crear_prestamo(999, miembro.id);
    assert_eq!(err, Err(ErrorBiblioteca::EntidadNoEncontrada));
}

#[test]
fn devolver_prestamo() {
    let mut s = setup();
    let libro = s
        .alta_libro("El Quijote", "Cervantes", "1", 1605, GeneroLibro::Ficcion)
        .unwrap();
    let miembro = s.alta_miembro("Ana", "ana@correo.com", "1").unwrap();
    let p = s.crear_prestamo(libro.id, miembro.id).unwrap();

    let devuelto = s.devolver_prestamo(p.id).expect("DevolverPrestamo inesperado");
    assert!(devuelto.devuelto, "el préstamo debe quedar devuelto");

    let de_nuevo = s.crear_prestamo(libro.id, miembro.id);
    assert!(de_nuevo.is_ok(), "tras la devolución el libro debe poder prestarse");

    let inexistente = s.devolver_prestamo(999);
    assert_eq!(inexistente, Err(ErrorBiblioteca::EntidadNoEncontrada));
}

#[test]
fn resumen() {
    let mut s = setup();
    s.alta_libro("A", "Autor", "1", 2020, GeneroLibro::Otro).unwrap();
    s.alta_libro("B", "Autor", "2", 2020, GeneroLibro::Otro).unwrap();
    let m = s.alta_miembro("Ana", "ana@correo.com", "1").unwrap();
    s.crear_prestamo(1, m.id).unwrap();

    let r = ReportesService::nuevo(&s);
    let resumen = r.resumen();
    assert_eq!(resumen.get("libros"), Some(&2));
    assert_eq!(resumen.get("libros_disponibles"), Some(&1));
    assert_eq!(resumen.get("libros_prestados"), Some(&1));
    assert_eq!(resumen.get("miembros_activos"), Some(&1));
    assert_eq!(resumen.get("prestamos_activos"), Some(&1));
}

#[test]
fn top_libros() {
    let mut s = setup();
    let libro_a = s.alta_libro("A", "Autor", "1", 2020, GeneroLibro::Otro).unwrap();
    let libro_b = s.alta_libro("B", "Autor", "2", 2020, GeneroLibro::Otro).unwrap();
    let m = s.alta_miembro("Ana", "ana@correo.com", "1").unwrap();

    let p1 = s.crear_prestamo(libro_a.id, m.id).unwrap();
    s.devolver_prestamo(p1.id).unwrap();
    let p2 = s.crear_prestamo(libro_a.id, m.id).unwrap();
    s.devolver_prestamo(p2.id).unwrap();
    let p3 = s.crear_prestamo(libro_b.id, m.id).unwrap();
    let _ = p3;

    let r = ReportesService::nuevo(&s);
    let top = r.top_libros(2);
    assert_eq!(top.len(), 2);
    assert_eq!(top[0].nombre, "A");
    assert_eq!(top[0].prestamos, 2);
    assert_eq!(top[1].nombre, "B");
    assert_eq!(top[1].prestamos, 1);
}

#[test]
fn top_miembros() {
    let mut s = setup();
    let m1 = s.alta_miembro("Ana", "ana@correo.com", "1").unwrap();
    let m2 = s.alta_miembro("Luis", "luis@correo.com", "2").unwrap();
    let libro_a = s.alta_libro("A", "Autor", "1", 2020, GeneroLibro::Otro).unwrap();
    let libro_b = s.alta_libro("B", "Autor", "2", 2020, GeneroLibro::Otro).unwrap();

    s.crear_prestamo(libro_a.id, m1.id).unwrap();
    s.crear_prestamo(libro_b.id, m1.id).unwrap();

    // m2 nunca pide prestado; aparece con 0 préstamos en el ranking.
    let _ = (m2.nombre.clone(), m2.id);

    let r = ReportesService::nuevo(&s);
    let top = r.top_miembros(2);
    assert_eq!(top.len(), 2);
    assert_eq!(top[0].nombre, "Ana");
    assert_eq!(top[0].prestamos, 2);
    assert_eq!(top[1].nombre, "Luis");
    assert_eq!(top[1].prestamos, 0);
}

#[test]
fn prestamos_vencidos() {
    let mut s = setup();
    let libro = s.alta_libro("A", "Autor", "1", 2020, GeneroLibro::Otro).unwrap();
    let m = s.alta_miembro("Ana", "ana@correo.com", "1").unwrap();
    s.crear_prestamo(libro.id, m.id).unwrap();

    let vencidos = s.prestamos_vencidos();
    assert!(vencidos.is_empty(), "un préstamo recién creado no está vencido");
}

#[test]
fn top_libros_descendente() {
    let mut s = setup();
    let libro_a = s.alta_libro("A", "Autor", "1", 2020, GeneroLibro::Otro).unwrap();
    let m = s.alta_miembro("Ana", "ana@correo.com", "1").unwrap();

    let p1 = s.crear_prestamo(libro_a.id, m.id).unwrap();
    s.devolver_prestamo(p1.id).unwrap();
    let p2 = s.crear_prestamo(libro_a.id, m.id).unwrap();
    let _ = p2;

    let r = ReportesService::nuevo(&s);
    let top = r.top_libros(1);
    assert_eq!(top.len(), 1);
    assert_eq!(top[0].prestamos, 2, "debe contar préstamos acumulados");
}

#[test]
fn buscar_sin_mayusculas() {
    let mut s = setup();
    s.alta_libro("Cien Años de Soledad", "Gabriel García Márquez", "2", 1967, GeneroLibro::Ficcion).unwrap();

    let res = s.buscar_libros("años");
    assert_eq!(res.len(), 1, "búsqueda en minúsculas debe encontrar 'Años'");
    let res = s.buscar_libros("GARCÍA");
    assert_eq!(res.len(), 1, "búsqueda en mayúsculas debe encontrar 'García'");
}

#[test]
fn ids_incrementales() {
    let mut s = setup();
    let libro1 = s.alta_libro("A", "Autor", "1", 2020, GeneroLibro::Otro).unwrap();
    let libro2 = s.alta_libro("B", "Autor", "2", 2020, GeneroLibro::Otro).unwrap();
    assert_eq!((libro1.id, libro2.id), (1, 2));
}

#[test]
fn ids_prestamos() {
    let mut s = setup();
    let libro_a = s.alta_libro("A", "Autor", "1", 2020, GeneroLibro::Otro).unwrap();
    let libro_b = s.alta_libro("B", "Autor", "2", 2020, GeneroLibro::Otro).unwrap();
    let m = s.alta_miembro("Ana", "ana@correo.com", "1").unwrap();

    let p1 = s.crear_prestamo(libro_a.id, m.id).unwrap();
    let p2 = s.crear_prestamo(libro_b.id, m.id).unwrap();
    assert_ne!(p1.id, p2.id, "los préstamos no deben compartir ID");
}